### DISCUSSION TITLE
Phlex 2.0 alpha is ready for testing 🏗️

### DISCUSSION THREAD

**Discussion by joeldrapper** [yippee-fun/phlex#769] [General]
The Phlex 2.0 alpha is ready for testing. New draft docs are in `main` here. There are instructions on how to build them in the `README`. Contributions to this documentation are very welcome. https://github.com/phlex-ruby/phlex.fun

You can install the alpha directly from GitHub.

```ruby
gem "phlex-rails", git: "https://github.com/phlex-ruby/phlex-rails.git"
gem "phlex", git: "https://github.com/phlex-ruby/phlex.git"
```

I recommend trying it out in a new Rails app. You’ll need Ruby `3.3.1` or greater. Try the generators, kick the tyres and let us know how you get on.

Kits are enabled by default with the new install generator, so you can render components by calling their name as a method, e.g.

```ruby
Card(title: "Hello") do
  p { "In a card" }
end
```

**Comment by joeldrapper** [comment]
You can now preview the new (incomplete) docs here. https://beta.phlex.fun

**Comment by stephannv** [comment]
I upgraded my app to Phlex 2.0.0.beta2 and only 3 things made my tests fail:

### Rendering translations with HTML
This code:
```ruby
# Before this rendered the correct html
div { t("my_key_html") }

# On Phlex 2.0.0.beta2 this escapes the HTML
div { t("my_key_html") }

# I had to do:
div { safe(t("my_key_html")) }
```

I'm not sure why, the `ActiveSupport::SafeBuffer.include(Phlex::SGML::SafeObject)` on Phlex::Rails should have resolved this. I will investigate this more.

### Rendering partials
```ruby
# before
render "my/partial/path"

#after
render partial: "my/partial/path"
```
I think this is fine, but just adding here because it could be added to the upgrading guide.

### Phlex symbols handling x Turbo symbols handling
I had:
```ruby
# on my controller
turbo_stream.replace(:x_y_z, XYZComponent.new) 

# on my view
div id: :x_y_z, ...
```
This rendered `<div id="x-y-z">...</div>` and turbo stream tried to replace a div with id `x_y_z`. I had to change to `div id: "x_y_z"`.

**Reply by joeldrapper** [reply]
Thank you.

1. The first issue with `t` is being sorted already. Should be fixed in the next beta.
2. The second issue is intentional because we want you to be able to render polymorphic inputs including strings. We need to document this well in the upgrade guide.
3. The third issue is also intentional, though this has led to much debate. 😅 I outlined my thoughts on the decision here. https://github.com/orgs/phlex-ruby/discussions/785

**Comment by mhenrixon** [comment]
Is this becuase of phlex 2 or because of something else?

Test:

```ruby
RSpec.describe Views::Organizations::Transfers::Show, type: :component do
  fixtures :organizations, :accounts, :transfers, :users, :members

  let(:organization) { organizations(:zazu) }
  let(:user)         { users(:alfred) }
  let(:member)       { members(:zazu_manager_alfred) }
  let(:transfer)     { transfers(:one) }
  let(:accounts)     { organization.accounts }

  it "displays the transfer details" do
    render described_class.new(
      member:,
      transfer:,
      accounts:
    )

    expect(component).to have_content(transfer.id)
    expect(component).to have_content(transfer.amount_currency)
    expect(component).to have_content(transfer.to_d.to_s)
    expect(component).to have_content(transfer.status)
    expect(component).to have_content(transfer.creditor_account.name)
    expect(component).to have_content(transfer.debitor_account.name)
    expect(component).to have_content(transfer.created_at.rfc3339)

    expect(component).to have_link("Back to Transfers", href: organization_transfers_path)
  end
end
```

Gives me the following error:

```
  1) Views::Organizations::Transfers::Show displays the transfer details
     Failure/Error:
       turbo_frame_tag :transfer_details do
         div(class: "py-6 px-6 bg-white h-screen h-full !w-[32rem]") do
           div(class: "space-y-6") do
             header
             payment_info
             transaction_details
             category_form
             labels_info
           end
         end

     NoMethodError:
       undefined method 'turbo_frame_tag' for nil
     # /Users/mhenrixon/.gem/ruby/3.4.0/gems/phlex-rails-2.0.0.beta2/lib/phlex/rails/helper_macros.rb:14:in 'Phlex::Rails::Helpers::TurboFrameTag#turbo_frame_tag'
     # ./app/views/organizations/transfers/show.rb:17:in 'Views::Organizations::Transfers::Show#view_template'
     # /Users/mhenrixon/.gem/ruby/3.4.0/gems/phlex-2.0.0.beta2/lib/phlex/sgml.rb:130:in 'block (2 levels) in Phlex::SGML#call'
     # /Users/mhenrixon/.gem/ruby/3.4.0/gems/phlex-2.0.0.beta2/lib/phlex/sgml.rb:319:in 'Phlex::SGML#around_template'
     # /Users/mhenrixon/.gem/ruby/3.4.0/gems/phlex-2.0.0.beta2/lib/phlex/sgml.rb:115:in 'block in Phlex::SGML#call'
     # /Users/mhenrixon/.gem/ruby/3.4.0/gems/phlex-2.0.0.beta2/lib/phlex/context.rb:23:in 'Phlex::Context#around_render'
     # /Users/mhenrixon/.gem/ruby/3.4.0/gems/phlex-2.0.0.beta2/lib/phlex/sgml.rb:114:in 'Phlex::SGML#call'
     # /Users/mhenrixon/.gem/ruby/3.4.0/gems/phlex-2.0.0.beta2/lib/phlex/testing/sgml.rb:5:in 'Phlex::Testing::SGML#render_to_string'
     # /Users/mhenrixon/.gem/ruby/3.4.0/gems/phlex-2.0.0.beta2/lib/phlex/testing/capybara.rb:25:in 'Phlex::Testing::Capybara#render'
     # ./spec/views/organizations/transfers/show_spec.rb:13:in 'block (2 levels) in <main>'
```

Not sure where to start looking

**Reply by stephannv** [reply]
> Is this becuase of phlex 2 or because of something else?
> 
> Test:
> 
> ```ruby
> RSpec.describe Views::Organizations::Transfers::Show, type: :component do
>   fixtures :organizations, :accounts, :transfers, :users, :members
> 
>   let(:organization) { organizations(:zazu) }
>   let(:user)         { users(:alfred) }
>   let(:member)       { members(:zazu_manager_alfred) }
>   let(:transfer)     { transfers(:one) }
>   let(:accounts)     { organization.accounts }
> 
>   it "displays the transfer details" do
>     render described_class.new(
>       member:,
>       transfer:,
>       accounts:
>     )
> 
>     expect(component).to have_content(transfer.id)
>     expect(component).to have_content(transfer.amount_currency)
>     expect(component).to have_content(transfer.to_d.to_s)
>     expect(component).to have_content(transfer.status)
>     expect(component).to have_content(transfer.creditor_account.name)
>     expect(component).to have_content(transfer.debitor_account.name)
>     expect(component).to have_content(transfer.created_at.rfc3339)
> 
>     expect(component).to have_link("Back to Transfers", href: organization_transfers_path)
>   end
> end
> ```
> 
> Gives me the following error:
> 
> ```
>   1) Views::Organizations::Transfers::Show displays the transfer details
>      Failure/Error:
>        turbo_frame_tag :transfer_details do
>          div(class: "py-6 px-6 bg-white h-screen h-full !w-[32rem]") do
>            div(class: "space-y-6") do
>              header
>              payment_info
>              transaction_details
>              category_form
>              labels_info
>            end
>          end
> 
>      NoMethodError:
>        undefined method 'turbo_frame_tag' for nil
>      # /Users/mhenrixon/.gem/ruby/3.4.0/gems/phlex-rails-2.0.0.beta2/lib/phlex/rails/helper_macros.rb:14:in 'Phlex::Rails::Helpers::TurboFrameTag#turbo_frame_tag'
>      # ./app/views/organizations/transfers/show.rb:17:in 'Views::Organizations::Transfers::Show#view_template'
>      # /Users/mhenrixon/.gem/ruby/3.4.0/gems/phlex-2.0.0.beta2/lib/phlex/sgml.rb:130:in 'block (2 levels) in Phlex::SGML#call'
>      # /Users/mhenrixon/.gem/ruby/3.4.0/gems/phlex-2.0.0.beta2/lib/phlex/sgml.rb:319:in 'Phlex::SGML#around_template'
>      # /Users/mhenrixon/.gem/ruby/3.4.0/gems/phlex-2.0.0.beta2/lib/phlex/sgml.rb:115:in 'block in Phlex::SGML#call'
>      # /Users/mhenrixon/.gem/ruby/3.4.0/gems/phlex-2.0.0.beta2/lib/phlex/context.rb:23:in 'Phlex::Context#around_render'
>      # /Users/mhenrixon/.gem/ruby/3.4.0/gems/phlex-2.0.0.beta2/lib/phlex/sgml.rb:114:in 'Phlex::SGML#call'
>      # /Users/mhenrixon/.gem/ruby/3.4.0/gems/phlex-2.0.0.beta2/lib/phlex/testing/sgml.rb:5:in 'Phlex::Testing::SGML#render_to_string'
>      # /Users/mhenrixon/.gem/ruby/3.4.0/gems/phlex-2.0.0.beta2/lib/phlex/testing/capybara.rb:25:in 'Phlex::Testing::Capybara#render'
>      # ./spec/views/organizations/transfers/show_spec.rb:13:in 'block (2 levels) in <main>'
> ```
> 
> Not sure where to start looking

Maybe it is a problem with your test helpers. https://beta.phlex.fun/reference/rails/testing.html

**Reply by mhenrixon** [reply]
No, that part is correct: `require "phlex/rails/testing"` and then a before on type: :component to include the damn thing.

**Reply by stephannv** [reply]
> No, that part is correct: `require "phlex/rails/testing"` and then a before on type: :component to include the damn thing.

Well, it looks like the problem is that your view_context is nil, and the phlex/rails/testing is responsible for defining view_context:
```ruby
def view_context
  controller.view_context
end

def controller
  @controller ||= ActionView::TestCase::TestController.new
end
```

You can try this to check if the setup is correct:
```ruby
it "..." do
  debugger # call `view_context`, `controller` and `controller.view_context` to check if these values are present
  # OR
  puts view_context
  puts controller
  puts controller.view_context
  
  render described_class.new(
    member:,
    transfer:,
    accounts:
  )
  ...
end
```

**Reply by stephannv** [reply]
You can ensure that `Phlex::Testing::Capybara` is included too.
https://github.com/phlex-ruby/phlex/blob/main/lib/phlex/testing/capybara.rb

I'm using Phlex 2 but I'm not using its test helpers, just doing this is working for me:
```ruby
# spec/support/phlex.rb
RSpec.configure do |config|
  %w[
    spec/components/
    spec/pages/
  ].each do |path|
    config.define_derived_metadata(file_path: Regexp.compile(path)) { |metadata| metadata[:type] ||= :view }
  end
end

# spec/rails_helper.rb
# uncomment this line
Rails.root.glob("spec/support/**/*.rb").each { |f| require f }
```
So, changing `type: :component` to `type: view` could work.

**Reply by mhenrixon** [reply]
Seems like a timing issue:

![CleanShot 2024-10-28 at 13 24 00@2x](https://github.com/user-attachments/assets/dacf0f44-7c2d-40d3-9ce8-25dc3ae75b58)


When I have the debugger statement here, the test passes even though the view context is nil, if I remove the debugger statement the test fails.

**Reply by joeldrapper** [reply]
I don’t know where `type: :view` is coming from. Phlex doesn’t define this, though we have documented how to set up a `type: :component` test yourself. https://beta.phlex.fun/reference/testing/capybara.html#defining-an-rspec-spec-type

**Reply by joeldrapper** [reply]
As @stephannv mentioned, it looks like the problem is the component isn’t being given a view_context. That is meant to be defined by this method https://github.com/phlex-ruby/phlex-rails/blob/80388e7c1a5ba53f11e8de45eeb250bb99bc9060/lib/phlex/rails/testing.rb#L7-L9

You should try `puts view_context` to see if it is indeed nil. You can also try `puts self.class.ancestors` and `puts method(:view_context).owner`. The owner should be `Phlex::Testing::RailsContext`.

**Reply by mhenrixon** [reply]
Huh... I don't get this:

```ruby
(ruby) puts method(:view_context).owner
Phlex::Testing::SGML
```

I am not requiring phlex/testing anywwhere and I require phlex/rails/testing at the top of everything. Why isn't that working?

**Reply by mhenrixon** [reply]
I also don't understand why putting `debugger` before the render and waiting a few seconds make the test pass fine @joeldrapper

**Reply by joeldrapper** [reply]
Okay, this is great progress. So we know the issue is that `view_context` here is coming from `Phlex::Testing::SGML` instead of `Phlex::Testing::RailsContext`. Now we need to figure out why. In theory, requiring `"phlex/rails/testing"` instead of `"phlex/testing"` should mean that method gets overridden.

There seems to be some kind of race condition that means the test is executed before that happens. Are you using a parallel or threaded test runner? RSpec isn’t parallel or threaded by default as far as I know.

**Reply by joeldrapper** [reply]
Also, can you confirm which version of Ruby you’re using? Especially relevant if it’s JRuby or TruffleRuby.

**Reply by stephannv** [reply]
> I don’t know where `type: :view` is coming from. Phlex doesn’t define this, though we have documented how to set up a `type: :component` test yourself. https://beta.phlex.fun/reference/testing/capybara.html#defining-an-rspec-spec-type

@joeldrapper using `type: :view`, you get the native [rspec-rails render method](https://github.com/rspec/rspec-rails/blob/main/lib/rspec/rails/example/view_example_group.rb#L66). This is why I didn't need to add phlex test helpers. 

Alternative: 
```ruby
RSpec.configure do |config|
  config.include RSpec::Rails::ViewExampleGroup, type: %i[component page]
end
```

**Reply by mhenrixon** [reply]
> Also, can you confirm which version of Ruby you’re using? Especially relevant if it’s JRuby or TruffleRuby.
Using ruby 3.4.0-preview because YOLO (naah seriously, we needed it for some phlex component related stuff actually which allows us to inherit from PhlexyUI components and customize them wihout crashing like in ruby 3.3.5)

> There seems to be some kind of race condition that means the test is executed before that happens. Are you using a parallel or threaded test runner? RSpec isn’t parallel or threaded by default as far as I know.

Nothing threaded or parallel in this test suite but we do have some serious performance issues with our test suite. Things that should be fast are painstakingly slow. Could this be related perhaps?

Even if I do the following manually:

```ruby
require "phlex/rails"
require "phlex/testing"

module Phlex::Testing::RailsContext
	def view_context
		controller.view_context
	end

	def controller
		@controller ||= ActionView::TestCase::TestController.new
	end
end

Phlex::Testing::SGML.include(
	Phlex::Testing::RailsContext,
)
```

In the test, it still fails

**Reply by stephannv** [reply]
@mhenrixon you can try removing all phlex testing helpers requires & includes, and try the snippet below:
```ruby
RSpec.describe Views::Organizations::Transfers::Show do
  include RSpec::Rails::ViewExampleGroup
  include Capybara::RSpecMatchers
  
  it "..." do
    render described_class.new(...)
  end
end
```

If it works, it is a problem with your Phlex testing setup. If not, it can be a problem with you spec/app/test setup.

**Reply by mhenrixon** [reply]
That worked as far as that I got a much deeper issue with url generation that I might have to solve but it at least doesn't crash at the missing turbo frame.

**Reply by stephannv** [reply]
> That worked as far as that I got a much deeper issue with url generation that I might have to solve but it at least doesn't crash at the missing turbo frame.

You can include `Rails.application.routes.url_helpers`. Something like:

```ruby
# spec/support/components.rb
RSpec.configure do |config|
  config.include RSpec::Rails::ViewExampleGroup, type: %i[component page] # or just :component if you don't need type: :page
  config.include Capybara::RSpecMatchers, type: %i[component page]
  config.include Rails.application.routes.url_helpers, type: %i[component page]
end
```

Or just using `type: :view` and `config.include Capybara::RSpecMatchers, type: :view`

**Reply by mhenrixon** [reply]
Yup! The following works like a charm:

```ruby
# TODO: Figure out how to use phlex testing to avoid system test everything
require "rails_helper"

RSpec.describe Views::Organizations::Transfers::Show, type: :view do
  fixtures :organizations, :accounts, :transfers, :users, :members

  let(:organization) { organizations(:zazu) }
  let(:user)         { users(:alfred) }
  let(:member)       { members(:zazu_manager_alfred) }
  let(:transfer)     { transfers(:one) }
  let(:accounts)     { organization.accounts }

  it "displays the transfer details" do
    render described_class.new(
      member:,
      transfer:,
      accounts:
    )

    expect(component).to have_content(transfer.id)
    expect(component).to have_content(transfer.amount_currency)
    expect(component).to have_content(transfer.to_d.to_s)
    expect(component).to have_content(transfer.status)
    expect(component).to have_content(transfer.creditor_account.name)
    expect(component).to have_content(transfer.debitor_account.name)
    expect(component).to have_content(transfer.created_at.rfc3339)

    expect(component).to have_link("Back to Transfers", href: organization_transfers_path)
  end

  def component
    response
  end
end
``

```ruby
# spec/rails_helper.rb
RSpec.configure do |config|
  config.include ActiveSupport::Testing::TimeHelpers # to use freeze_time and travel_to
  config.include RSpec::Rails::ViewExampleGroup, type: %i[component page] # or just :component if you don't need type: :page
  config.include Capybara::RSpecMatchers, type: %i[component page view]
  config.include Rails.application.routes.url_helpers, type: %i[component page view]
end
```

Maybe you can have a look at that @joeldrapper seems like it shouldn't be this hard. I believe something is loading out of order.

**Reply by stephannv** [reply]
> Yup! The following works like a charm:
> 
> ```ruby
> # TODO: Figure out how to use phlex testing to avoid system test everything
> require "rails_helper"
> 
> RSpec.describe Views::Organizations::Transfers::Show, type: :view do
>   fixtures :organizations, :accounts, :transfers, :users, :members
> 
>   let(:organization) { organizations(:zazu) }
>   let(:user)         { users(:alfred) }
>   let(:member)       { members(:zazu_manager_alfred) }
>   let(:transfer)     { transfers(:one) }
>   let(:accounts)     { organization.accounts }
> 
>   it "displays the transfer details" do
>     render described_class.new(
>       member:,
>       transfer:,
>       accounts:
>     )
> 
>     expect(component).to have_content(transfer.id)
>     expect(component).to have_content(transfer.amount_currency)
>     expect(component).to have_content(transfer.to_d.to_s)
>     expect(component).to have_content(transfer.status)
>     expect(component).to have_content(transfer.creditor_account.name)
>     expect(component).to have_content(transfer.debitor_account.name)
>     expect(component).to have_content(transfer.created_at.rfc3339)
> 
>     expect(component).to have_link("Back to Transfers", href: organization_transfers_path)
>   end
> 
>   def component
>     response
>   end
> end
> ``
> 
> ```ruby
> # spec/rails_helper.rb
> RSpec.configure do |config|
>   config.include ActiveSupport::Testing::TimeHelpers # to use freeze_time and travel_to
>   config.include RSpec::Rails::ViewExampleGroup, type: %i[component page] # or just :component if you don't need type: :page
>   config.include Capybara::RSpecMatchers, type: %i[component page view]
>   config.include Rails.application.routes.url_helpers, type: %i[component page view]
> end
> ```
> 
> Maybe you can have a look at that @joeldrapper seems like it shouldn't be this hard. I believe something is loading out of order.

If you don't want to create the `component` method, you can do something like:
```ruby
component = render MyClass.new

expect(component).to ...
```
